Support for OrganizationIdentifier and PSD2 lints

mtgag opened the pull request:
Hello,

in this fork we have added some lints related to the OrganizationIdentifier field and PSD2. Please consider integrating them into the project.

zakird commented:
Thanks for the pull request. I sincerely appreciate the code to check conformance with the ETSI documents.

I've had a chance to discuss this PR with several ZLint maintainers. None of us know the ETSI requirements well enough to review these lints to the quality expected of ZLint. The current maintainers also don't follow the ETSI standards closely enough to make sure that ZLint stays inline with the requirements as they evolve. In contrast, several of us are involved with the CA/Browser Forum and/or operate CAs that conforms with those standards. This ensures that ZLint will mirror those documents moving forward.

As such, we don't feel that we can merge this PR into ZLint. We recommend creating a separate repository of lints for the ETSI documents. Given that ZLint expects a standard format, you should be able to use the lints with the ZLint framework fairly seamlessly despite their maintenance in a separate repo.

sleevi commented:
Zakir: Is there anything I can do to help? Possibly with the assistance of
Wayne Thayer of Mozilla?

From the browser side, I think this would be hugely valuable to have as a
part of zlint, to the extent browsers still accept ETSI and ETSI still
requires TLS, and would love to figure how to help if that’s something that
works for you and the other maintainers.

cardonator commented:
@zakird I think your comment suggests a somewhat different conversation around how we could better group lints. Zlint currently has over 250 lints that are not grouped in any particular way in the code. Does it make sense to begin to break these out into the groups of standards the lints apply to? Maybe a different form of grouping could lend better to having different groups of maintainers over different sets of lints?

zakird commented:
Hi all, thanks for your patience on this issue. @cardonator and @sleevi, your perspectives on this are helpful. I've had a chance to touch base with other maintainers and we'd like to do the following:

1) Split up the lints by standard into separate directories: RFCs, CA/B BRs, CA/B EV Requirements, Mozilla Root Store, and ETSI. @cpu and I will be split these up in January. Longterm, this will accommodate having different code owners for each set of lints. We're going to temporarily hold off merging both this PR and the Mozilla one until this split takes place.

2) Extend the ZLint API to allow users (i.e., CAs and auditors) to select which sets of lints they want to run a certificate against. Not all sets of lints will be run by default, and we'll need to decide what sources are run by default, but ETSI won't initially be in the default set. There's an argument that only RFC should be default. We can discuss this separately. 

3) Our initial concern was that none of the core maintainers know the ETSI documents well enough to appropriately review these lints for policy correctness. Longterm, we need to see CAs help both contribute _and_ review ETSI lints for correctness. Until we're able to build a community, @sleevi has agreed to help be a code owner to ensure their policy correctness. The existing maintainers will review the lints to ensure ZLint code standards.

How does this sound to to folks?

cpu commented:
@mtgag Can you merge master into this branch? Most notably since https://github.com/zmap/zlint/pull/337 landed you'll find the ETSI related lints are all located in `lints/etsi/` now. Once this branch is caught up I'll schedule some time to do a general code review (waiting on the ETSI specifics to be reviewed by another subject matter expert).

Thanks!

mtgag commented:
Please excuse the delayed response and participation in this topic due
to the holiday period here in Germany. We initiated this lint after an
order from T-Systems, which uses zlint in their CA, and wanted to
contribute to this project by implementing and integrating these lints.
We will adress the issues regarding merging and spliting soon and write
on this topic.

cpu commented on lints/base.go:
```diff
@@ -54,7 +54,10 @@ const (
+	EtsiEsi                         // ETSI EN 319 412-5 V2.1.1 (ETSI - Electronic Signatures and Infrastructures (ESI))
+	EtsiTs_119_495_EsiPsd           // ETSI TS 119 495 V1.2.1 ("PSD2 Directive")
+	EtsiEn_319_412_1_CommDataStruct // ETSI EN 319 412-1 Overview and Common Data Structures
+	EtsiEn_319_412_4_CertProfileWeb // ETSI EN 319 412-4
```
@sleevi I agree with your order of preferences. I think using `EtsiEsi` for anything related to these profiles would be fine. Thanks for the additional background. It definitely improved my understanding of the broader context of these lints.

mtgag replied:
We implemented the EtsiEsi solution to resolve this issue.

cpu commented on lints/lint_ev_ntr_subject_jurisdiction_serial.go:
```diff
@@ -0,0 +1,99 @@
+/*
```
> @cpu Question: Should this lint execute if it is not well-formed according to 9.2.8 of the EVGs? For example, if the field is only 4 characters, it's not syntactically well-formed. If there is no 'hyphen-minus' character, it's not syntactically well formed. Either of those cases will cause subsequent checks to fail, but it seems that they belong as a separate lint, and this lint should only execute if things are "close enough" to being well-formed.

I don't have a strong preference for this case but in general smaller lints with tighter focus are easier to review. Unless @mtgag feels that it makes the lint results harder to understand I'd be in favour of splitting off the well-formed checks on the basis of making the review process easier.

mtgag replied:
No strong preference here either. If the name of the lint is "expressive" enough, smaller lints seem all right.

mtgag replied:
We have splitted into two lints:
lint_ev_orgid_well_fomed.go and lint_ev_ntr_subject_jurisdiction_serial.go

cpu commented:
Thanks for the update!

weddi-eddy commented:
> Split up the lints by standard into separate directories: RFCs, CA/B BRs, CA/B EV Requirements, Mozilla Root Store, and ETSI. @cpu and I will be split these up in January.

Hrmm.. I did notice that with a previous release and this explains it. I think it's a great idea and gave me one of my own. Would it make sense to add an **experimentals** folder for lints such as this one that are suggested but need additional testing and comment? I guess by default, these should be excluded in the API, but can be included for CA and auditor development environments. It gives internal devs an easier to way to test it and possibly provide feedback.

Hrmm.. I suppose that if you did this however, you would need to think about how experimental lints are stored.  If you had 100 suggested lints for testing, not everyone would want to download them in a production environment, so some thought would need to go into handling that.

I still think it's a good idea to consider. It gives a specific folder for certificate testers, CAs and potential lint writers to play in without messing with the official folders.

cpu commented:
>  Would it make sense to add an experimentals folder for lints such as this one that are suggested but need additional testing and comment? I guess by default, these should be excluded in the API, but can be included for CA and auditor development environments. It gives internal devs an easier to way to test it and possibly provide feedback.

@weddi-eddy I think you're imagining something closer to a plugin system and I'm not sure the complexity is justified. With the existing codebase that assumes lints are compiled-in I believe any lints that need additional testing and comment should have that done during the pull request stage, before they are merged to master.

> I still think it's a good idea to consider. It gives a specific folder for certificate testers, CAs and potential lint writers to play in without messing with the official folders.

If you're interested in this feature request could you open a new issue for it? I think that would be a better place to discuss than on this PR.

Thanks!

cpu commented:
:wave: @mtgag I saw you recently pushed some commits to this branch addressing review feedback. Could you update the feedback from @sleevi that you addressed with comments/marking it resolved? Do you think this branch is ready for a re-review?

mtgag commented:
> 
> 
> 👋 @mtgag I saw you recently pushed some commits to this branch addressing review feedback. Could you update the feedback from @sleevi that you addressed with comments/marking it resolved? Do you think this branch is ready for a re-review?

we are still working on it. I hope we will be finished soon and we will update this topic.

mtgag commented on lints/lint_ev_ntr_subject_jurisdiction_serial.go:
```diff
@@ -0,0 +1,99 @@
+
+	subjOrgId := util.GetSubjectOrgId(c.RawSubject)
+	if util.IsExtInCert(c, util.CabfExtensionOrganizationIdentifier) {
+		_, result = util.ParseCabfOrgIdExt(c) // check error string during Execute
```
Added an extra test in Execute.

mtgag commented on lints/lint_ev_orgid_encoding.go:
```diff
@@ -0,0 +1,63 @@
+func (l *evOrgIdEncoding) Execute(c *x509.Certificate) *LintResult {
+	orgId := util.GetSubjectOrgId(c.RawSubject)
+	var errStr string
+	reencParams := []string{"printable", "utf8"}
```
Implemented the proposed approach.

mtgag commented on lints/lint_ev_orgidext_matches_subject.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+func (l *evOrgIdExtMatchesSubject) Execute(c *x509.Certificate) *LintResult {
+	orgId := util.GetSubjectOrgId(c.RawSubject)
```
Of OI is found two times the lints fail. Test certificates provided.

mtgag commented on lints/lint_ev_orgid.go:
```diff
@@ -0,0 +1,57 @@
+	errStr, _ := util.ParseCabfOrgId(orgId.Value)
+	if errStr != "" {
+		return &LintResult{Status: Error, Details: errStr}
+	}
```
Separated into two different lints lint_ev_orgid.go and lint_ev_orgid_etsi.go. Test certificates provided.

mtgag commented on lints/lint_ev_ntr_subject_jurisdiction_serial.go:
```diff
@@ -0,0 +1,99 @@
+	return nil
+}
+
+func GetOrgIdFromSubjOrExt(c *x509.Certificate) (bool, util.ParsedEvOrgId) {
```
Copied this review comment verbatim into the code.

mtgag commented on lints/lint_ev_orgid.go:
```diff
@@ -0,0 +1,57 @@
+	_, isPsd2 := util.IsQcStatemPresent(c, &util.IdEtsiPsd2Statem)
+	orgId := util.GetSubjectOrgId(c.RawSubject)
+	if isPsd2 || !util.IsEV(c.PolicyIdentifiers) || !orgId.IsPresent {
+		// PSD2 certificates follow different requirements
```
This lint has been splitted into two different lints for standard EV and ETSI. For Standard EV the check isPsd2 is removed.

mtgag commented on lints/lint_ev_orgidext_matches_subject.go:
```diff
@@ -0,0 +1,59 @@
+	}
+	_, parsedOrgId := util.ParseCabfOrgId(orgId.Value)
+	// no need to check an error parsing the subject:organizationIdentifier here, this is done in other lints dealing with that field explicitly.
+	// StateOrProvince doesn't have to match literally, only semantically
```
While implementing this lint we discussed this point intensively. CAB Forum requires a 2 character ISO 3166-2 identifier for the subdivision (state or province) in subject:organizationIdentifier (Sec. 9.2.8) but the extension cabfOrganizationIdentifier (Sec. 9.8.2) an up to 128-character long registrationStateOrProvince. We thought the intention was to have CA in the OI field but allow California in the extension (see example above). If these two fields must really be equal then the registrationStateOrProvince in the extension should be declared with SIZE(2).

We will resolve this is issue like that: If subject:organizationIdentifier has state or province, then registrationStateOrProvince is not allowed to be empty. Ok?

mtgag commented:
We believe that we resolved the issues of the review. @sleevi Could you please check if all issues are properly adressed? If everything is fine we will proceed with the merging.

sleevi commented:
@mtgag I'm about to take a look (apologies, I was out of the office), but could you clarify https://github.com/zmap/zlint/pull/336#issuecomment-579108834

One of the suggestions from https://github.com/zmap/zlint/pull/336#pullrequestreview-338169149 was to try to reduce the size of the pull request to managable, independent chunks, which make it easier to review and make progress on. Is that a possibility here?

mtgag commented:
> 
> 
> @mtgag I'm about to take a look (apologies, I was out of the office), but could you clarify [#336 (comment)](https://github.com/zmap/zlint/pull/336#issuecomment-579108834)
> 
> One of the suggestions from [#336 (review)](https://github.com/zmap/zlint/pull/336#pullrequestreview-338169149) was to try to reduce the size of the pull request to managable, independent chunks, which make it easier to review and make progress on. Is that a possibility here?

Yes. Is it ok if we synchronise to master (hard reset) and slowly commit the lints once again?

sleevi commented:
> Yes. Is it ok if we synchronise to master (hard reset) and slowly commit the lints once again?

I don't see any reason why that'd be problematic.

Certainly, use a workflow that works for you, but if I was doing this (trying to break a very large branch into smaller PRs), my flow would likely be:

- Merge in `master` and make sure my lints are working
- `git rebase -i` to squash into a single commit that has all the lints
- `git rebase -i` to edit that single commit and
  - selectively, `git add` and `git commit` so that each logical lint is a commit
- At this point, you will have a single branch, with likely 5+ commits, each representing a logical lint (or set of lints, for a particular requirement)
- `git log` to view those commits, then `git checkout -b ${new_branch_foo/bar/baz/bat} ${hashA}`, for each hash in the set A...E, to create a new branch foo/bar/baz/bat/etc containing the state up to that particular hash
- `git rebase -i master` to rebase `${new_branch_foo}` onto master. This time, I would drop every commit other than the particular commit that I'm wanting to create a PR for

The end result of all of these shenanigans is that you'll have several branches, each with a single commit difference compared to `master`, each with a logical set of lints. Each of these can be opened as PRs, reviewed independently, edited independently.

The handwave here is what represents a discrete set of lints. https://github.com/zmap/zlint/pull/336#pullrequestreview-338169149 tried to outline some possible chunks, and which likely would not overlap in the same area (e.g. both needing to touch a utilities file).

As I mentioned, you're totally free to do it how you see fit! As long as it's possible to review smaller, discrete chunks that focus on solving a single logical set of requirements, I'm more than happy. There should be nothing wrong with a bunch of smaller CLs, and makes it easier to quickly review and approve and iterate :)

mtgag commented:
> 
> 
> > Yes. Is it ok if we synchronise to master (hard reset) and slowly commit the lints once again?
> 
> I don't see any reason why that'd be problematic.
> 
> Certainly, use a workflow that works for you, but if I was doing this (trying to break a very large branch into smaller PRs), my flow would likely be:
> 
>     * Merge in `master` and make sure my lints are working
> 
>     * `git rebase -i` to squash into a single commit that has all the lints
> 
>     * `git rebase -i` to edit that single commit and
>       
>       * selectively, `git add` and `git commit` so that each logical lint is a commit
> 
>     * At this point, you will have a single branch, with likely 5+ commits, each representing a logical lint (or set of lints, for a particular requirement)
> 
>     * `git log` to view those commits, then `git checkout -b ${new_branch_foo/bar/baz/bat} ${hashA}`, for each hash in the set A...E, to create a new branch foo/bar/baz/bat/etc containing the state up to that particular hash
> 
>     * `git rebase -i master` to rebase `${new_branch_foo}` onto master. This time, I would drop every commit other than the particular commit that I'm wanting to create a PR for
> 
> 
> The end result of all of these shenanigans is that you'll have several branches, each with a single commit difference compared to `master`, each with a logical set of lints. Each of these can be opened as PRs, reviewed independently, edited independently.
> 
> The handwave here is what represents a discrete set of lints. [#336 (review)](https://github.com/zmap/zlint/pull/336#pullrequestreview-338169149) tried to outline some possible chunks, and which likely would not overlap in the same area (e.g. both needing to touch a utilities file).
> 
> As I mentioned, you're totally free to do it how you see fit! As long as it's possible to review smaller, discrete chunks that focus on solving a single logical set of requirements, I'm more than happy. There should be nothing wrong with a bunch of smaller CLs, and makes it easier to quickly review and approve and iterate :)

cpu commented:
@mtgag It looks like you were able to open a handful of smaller PRs (Thanks!), can this larger PR be closed now?

mtgag commented:
Yes. The content of this large PR is contained in smaller ones.

sleevi commented:
Thanks! Sorry for the delay in getting to them, this week has been super
busy, but I’m hoping to start reviewing in full this weekend and leading
through next week. I really appreciate you doing this.
